### PR TITLE
JPA map storage: Authentication session no-downtime store

### LINKED ISSUE #9665: JPA map storage: Authentication session no-downtime store
### Description

Create a JPA map storage implementation for Authentication session area

### PR DESCRIPTION
**Pull request by vramik**
Closes #9665

### CODE REVIEW

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authSession/entity/JpaRootAuthenticationSessionEntity.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+/**
+ * There are some fields marked by {@code @Column(insertable = false, updatable = false)}.
+ * Those fields are automatically generated by database from json field,
+ * therefore marked as non-insertable and non-updatable to instruct hibernate.
+ */
+@Entity
+@Table(name = "kc_auth_root_session")
+@TypeDefs({@TypeDef(name = "jsonb", typeClass = JsonbType.class)})
+public class JpaRootAuthenticationSessionEntity extends AbstractRootAuthenticationSessionEntity implements JpaRootEntity {
```
Comment (ahus1) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authSession/entity/JpaRootAuthenticationSessionEntity.java:53]: I remember that our tests showed that the AuthenticationSessionEntity doesn't allow for optimistic locking as it is modified in two different transactions. I gave it a short try to see if the RootAuthenticationSessionEntity allows for a versionID, and a short test (log in via the web console) showed me that this worked. What do you think? Did you give it a try?

File: model/map-jpa/src/main/resources/META-INF/auth-sessions/jpa-auth-sessions-changelog-1.xml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            <column name="id" type="UUID">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="metadata" type="json"/>
+        </createTable>
+        <ext:addGeneratedColumn tableName="kc_auth_root_session">
+            <ext:column name="entityversion" type="INTEGER" jsonColumn="metadata" jsonProperty="entityVersion"/>
+            <ext:column name="realmid" type="VARCHAR(36)" jsonColumn="metadata" jsonProperty="fRealmId"/>
+            <ext:column name="timestamp" type="INTEGER" jsonColumn="metadata" jsonProperty="fTimestamp"/>
+            <ext:column name="expiration" type="BIGINT" jsonColumn="metadata" jsonProperty="fExpiration"/>
```
Comment (ahus1) [inline model/map-jpa/src/main/resources/META-INF/auth-sessions/jpa-auth-sessions-changelog-1.xml:38]: I see that "timestamp" is an integer (as it is also within Keycloak's logical layer), and there is "expiration" (with is AFAIK also a timestamp, but as a long). 

I wonder if we should treat both as long within the map storage (across all modules) to make this consistent. 

And if we should place it into the database as a timestamp, although I know having timestamps in databases has also its issues with regards to timezones.

**Review by ahus1 — CHANGES_REQUESTED** [review]
Thank for your this change, one more entity in JPA map storage!

File: model/map-jpa/src/main/resources/META-INF/auth-sessions/jpa-auth-sessions-changelog-1.xml (hunk 2 above)
Comment (hmlnarik) [reply model/map-jpa/src/main/resources/META-INF/auth-sessions/jpa-auth-sessions-changelog-1.xml:38]: > I wonder if we should treat both as long within the map storage (across all modules) to make this consistent.

Good catch. Yes, we should, could you please introduce a GHI for that?

> I also wonder if we should place it into the database as a timestamp, although I know having timestamps in databases has also its issues with regards to time zones.

I'd rather keep it as `BIGINT`, that corresponds to the usage of it within Keycloak. If a view of it should be needed by an admin, they can use DB functions for conversions in SELECTs

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authSession/entity/JpaRootAuthenticationSessionEntity.java (hunk 1 above)
Comment (vramik) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/authSession/entity/JpaRootAuthenticationSessionEntity.java]: I've tried it and it turned out we can make JpaRootAuthenticationSessionEntity implement JpaRootVersionedEntity.

File: model/map-jpa/src/main/resources/META-INF/auth-sessions/jpa-auth-sessions-changelog-1.xml (hunk 2 above)
Comment (ahus1) [reply model/map-jpa/src/main/resources/META-INF/auth-sessions/jpa-auth-sessions-changelog-1.xml:41]: New GHI: #10960 - beware of the Epochalypse!

As there is a new issue to handle this, this review comment is thereby resolved.

**Review by ahus1 — APPROVED** [review]

**Review by hmlnarik — APPROVED** [review]
Approving per @ahus1 's approval

**Pull request merged**